docs: Fix examples (fixes #720)

rasa opened the pull request:
I had to add this code to get these examples to work, under click-8.1.6, click-extra-4.6.3, and cloup-3.0.0.

kdeldycke commented:
Thanks @rasa for the PR, which is perfectly valid!

But I'm a bit worried by the additional forcing of the `package_name`. I'm sure you added them because there is no way to have the example work.

Unfortunately this uncover another, deeper issue: the package name autodetection doesn't work as intended in the context of the sphinx `.. click:example::` directives.

I'm too distracted right to dive deep into Click Extra code. But can you provide here the result of the output in the documentation, with and without this additional `package_name` parameter?

If you have enough courage, you can even dive into the magic of name detection at:

https://github.com/kdeldycke/click-extra/blob/05ac36e4f0a8cc5be9d95a96b4e4aafe99b416f0/click_extra/version.py#L131-L190

rasa commented:
Yeah, I actually wasn't expecting this PR to be accepted, but simply to open a discussion. I should have made that clear in my comment.

rasa commented:
> But can you provide here the result of the output in the documentation, with and without this additional `package_name` parameter?

With the `package_name` parameter, it works as expected (though the version is click-extra's version):
```text
python3 commands-06.py --version
commands-06.py, version 4.6.4
```
Without `package_name`, we get:
```text
  File ".../click-extra/click_extra/version.py", line 188, in package_name
    raise RuntimeError(msg)
RuntimeError: Could not determine the package name automatically from the frame stack. Try passing 'package_name' instead.
```
This patch works around the issue:
```diff
diff --git a/click_extra/version.py b/click_extra/version.py
index 1532d92..c476502 100644
--- a/click_extra/version.py
+++ b/click_extra/version.py
@@ -177,16 +177,6 @@ class ExtraVersionOption(ExtraOption):
         # https://docs.python.org/3/library/inspect.html#the-interpreter-stack
         del frame

-        if not package_name:
-            logger = logging.getLogger("click_extra")
-            for counter, (p_name, f_name) in enumerate(frame_chain):
-                logger.debug(f"Inspected frame #{counter}: {p_name}, {f_name}")
-            msg = (
-                "Could not determine the package name automatically from the frame "
-                "stack. Try passing 'package_name' instead."
-            )
-            raise RuntimeError(msg)
-
         return package_name

     @cached_property
@@ -197,6 +187,8 @@ class ExtraVersionOption(ExtraOption):
         <https://docs.python.org/3/library/importlib.metadata.html?highlight=metadata#distribution-versions>`_
         on the module whose ID is given by ``self.package_name``.
         """
+        if not self.package_name:
+            return "n/a"
         try:
             version = metadata.version(self.package_name)
         except metadata.PackageNotFoundError:
```

My guess is this is not the solution you're looking for, though....

rasa commented:
To expound on this:
https://github.com/kdeldycke/click-extra/blob/35fc4b55a34b1287f3829b7eb92c7babdb8965ad/click_extra/version.py#L143-174 :
```python
        for frame_info in inspect.stack():
            frame = frame_info.frame

            # Get the current package name from the frame's globals.
            frame_name = frame.f_globals["__name__"]

            # Keep track of the inspected frames.
            frame_chain.append((frame_name, frame_info.function))

            # Stop at the invoke() function of any CliRunner class, which is used for
            # testing.
            if frame_info.function == "invoke" and isinstance(
                frame.f_locals.get("self"),
                click.testing.CliRunner,
            ):
                pass

            # Skip the intermediate frames added by the `@cached_property` decorator
            # and the Click ecosystem.
            elif frame_name.startswith(("functools", "click_extra", "cloup", "click")):
                continue

            # We found the frame where the CLI is implemented.
            package_name = frame_name.split(".")[0]

            # Re-interpret the package name if it defined as `__main__` entry-point.
            if package_name == "__main__":
                package_name = frame.f_globals.get("__package__")
                if package_name:
                    package_name = package_name.split(".")[0]

            break
```
doesn't find anything in `__package__` in a simple script, such as the one found at
https://kdeldycke.github.io/click-extra/commands.html#remove-default-options :
```python
from click_extra import extra_command

@extra_command(params=None)
def bare_cli():
   pass

bare_cli()
```
So this fix would return the script name for the package name:
```diff
diff --git a/click_extra/version.py b/click_extra/version.py
index 1532d92..14fc6c6 100644
--- a/click_extra/version.py
+++ b/click_extra/version.py
@@ -170,6 +170,10 @@ class ExtraVersionOption(ExtraOption):
                 package_name = frame.f_globals.get("__package__")
                 if package_name:
                     package_name = package_name.split(".")[0]
+                    break
+                package_name = frame.f_globals.get("__file__")
+                if package_name:
+                    (_, package_name) = os.path.split(package_name)
 
             break
 
```
Is that the kind of fix you are looking for?

kdeldycke commented:
> Is that the kind of fix you are looking for?

Yeah, that sounds sensible to me. I think that's also the behaviour Click implements: to fallback on the script filename in that case. We should double check Click code but in the state of things, I'm OK with that trick. If you add that to the PR and confirm you don't need to force the package name as parameter, I'll merge it! :)

rasa commented:
> If you add that to the PR and confirm you don't need to force the package name as parameter, I'll merge it! :)

See #729 . The only issue left in this PR are the missing parans on a `my_cli.command` decorator.

kdeldycke commented:
Good catch by the way and thanks for the updated docs! Let's merge! :)
